Internal(Components): Prevent multiple users editing in the same time

MuliElementor opened the pull request:
- Added Document_Lock_Manager class to handle locking and unlocking of components.
- Introduced new REST API routes for locking, unlocking, and checking lock status of components.
- Integrated locking mechanism into the Components_REST_API class.
- Updated Module class to register lock-related hooks for component editing.
- Enhanced API client with methods for locking and unlocking components, and checking lock status.
- Added UI components to inform users when a document is being edited by another user.

This feature improves collaborative editing by preventing conflicts when multiple users attempt to edit the same component.

## PR Checklist
<!-- 
Please check if your PR fulfills the following requirements:
**Filling out the template is required.** Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
 -->
- [ ] The commit message follows our guidelines:  https://github.com/elementor/elementor/blob/master/.github/CONTRIBUTING.md


## PR Type
What kind of change does this PR introduce?
<!-- Please check the one that applies to this PR using "x" with no spaces eg: [x]. -->
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Summary

This PR can be summarized in the following changelog entry:

*

## Description
An explanation of what is done in this PR

*

## Test instructions
This PR can be tested by following these steps:

*

## Quality assurance

- [ ] I have tested this code to the best of my abilities
- [ ] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

Fixes #

<!--start_gitstream_placeholder-->
### ✨ PR Description
Purpose: Add document locking mechanism to prevent concurrent editing of the same component by multiple users

Main changes:
- Implemented Document_Lock_Manager class for managing component locks with user and timestamp tracking
- Added REST API endpoints for lock/unlock/status operations with appropriate permission checks
- Created UI dialog to notify users when a component is being edited by someone else
- Added client-side API integration to check lock status before switching to component editing

_Generated by LinearB AI and added by gitStream._
<sub>AI-generated content may contain inaccuracies. Please verify before using. **[We'd love your feedback!](mailto:product@linearb.io)** 🚀</sub>
<!--end_gitstream_placeholder-->

MuliElementor committed:
Implement document locking functionality in Components REST API

- Added Document_Lock_Manager class to handle locking and unlocking of components.
- Introduced new REST API routes for locking, unlocking, and checking lock status of components.
- Integrated locking mechanism into the Components_REST_API class.
- Updated Module class to register lock-related hooks for component editing.
- Enhanced API client with methods for locking and unlocking components, and checking lock status.
- Added UI components to inform users when a document is being edited by another user.

This feature improves collaborative editing by preventing conflicts when multiple users attempt to edit the same component.

MuliElementor committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time

MuliElementor committed:
format

MuliElementor committed:
format

elementorbot committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time

MuliElementor committed:
ךןמל

MuliElementor committed:
ךןמל

MuliElementor committed:
update

MuliElementor committed:
update

MuliElementor committed:
update

MuliElementor committed:
update

MuliElementor committed:
update

MuliElementor committed:
update

MiryamOren commented on modules/components/components-rest-api.php:
```diff
@@ -72,6 +82,77 @@ private function register_routes() {
+			[
+				'methods' => 'POST',
+				'callback' => fn( $request ) => $this->route_wrapper( fn() => $this->lock_component( $request ) ),
+				'permission_callback' => fn() => is_user_logged_in(),
```
```suggestion
				'permission_callback' => fn() => current_user_can( 'edit_posts' ),
```

MiryamOren commented on modules/components/components-rest-api.php:
```diff
@@ -72,6 +82,77 @@ private function register_routes() {
+			[
+				'methods' => 'POST',
+				'callback' => fn( $request ) => $this->route_wrapper( fn() => $this->unlock_component( $request ) ),
+				'permission_callback' => fn() => is_user_logged_in(),
```
```suggestion
				'permission_callback' => fn() => current_user_can( 'edit_posts' ),
```

MiryamOren commented on modules/components/document-lock-manager.php:
```diff
@@ -0,0 +1,97 @@
+			wp_set_post_lock( $component_id );
+
+			update_post_meta( $component_id, '_' . $this->document_type . '_lock_user', $user_id );
+			update_post_meta( $component_id, '_' . $this->document_type . '_lock_time', time() );
```
Why do we need the document_type?

MuliElementor committed:
Refactor: Simplify lock manager retrieval in Components_REST_API

- Replaced the existing lock manager initialization in Components_REST_API with a call to the new static method get_lock_manager_instance() in the Module class.
- Added get_lock_manager_instance() method to the Module class to provide a singleton instance of Document_Lock_Manager, ensuring consistent access across the application.

MuliElementor committed:
Refactor: Simplify lock manager retrieval in Components_REST_API

- Replaced the existing lock manager initialization in Components_REST_API with a call to the new static method get_lock_manager_instance() in the Module class.
- Added get_lock_manager_instance() method to the Module class to provide a singleton instance of Document_Lock_Manager, ensuring consistent access across the application.

MuliElementor committed:
update

MiryamOren commented on modules/components/components-rest-api.php:
```diff
@@ -72,6 +82,77 @@ private function register_routes() {
+				],
+			],
+		] );
+	}
```
let's merge these EPs, so we'll get all the info in one request, so the response will be something like
```
{
    is_current_user_allow_to_edit: boolean;
    locked_by: string | null;
}
` ``

MiryamOren commented on modules/components/components-rest-api.php:
```diff
@@ -72,6 +82,77 @@ private function register_routes() {
+			],
+		] );
+
+		register_rest_route( self::API_NAMESPACE, '/' . self::API_BASE . '/unlock', [
```
I don't think we should externalize this functionality via EP (and we don't use it)

MuliElementor committed:
Refactor: Streamline lock manager access in Module class

- Updated the get_lock_manager method to utilize the new static method get_lock_manager_instance, enhancing code clarity and consistency.
- Ensured that the lock manager is accessed through a singleton instance, improving resource management.

MuliElementor committed:
Refactor: Streamline lock manager access in Module class

- Updated the get_lock_manager method to utilize the new static method get_lock_manager_instance, enhancing code clarity and consistency.
- Ensured that the lock manager is accessed through a singleton instance, improving resource management.

MuliElementor committed:
כןס

MuliElementor committed:
כןס

MiryamOren commented on packages/packages/core/editor-components/src/create-component-type.ts:
```diff
@@ -74,23 +76,29 @@ function createComponentView( options: CreateTemplatedElementTypeOptions ): type
+		async switchDocument() {
+			const { isAllowedToSwitchDocument } = await canSwitchDocument(this.getComponentId()?.value as number);
+			if (!isAllowedToSwitchDocument) {
+				throw new Error('You are not allowed to switch this document');
```
Shouldn't we open the "this component is locked" popup?

MiryamOren commented on packages/packages/core/editor-components/src/components/in-edit-mode.tsx:
```diff
@@ -0,0 +1,43 @@
+import { __ } from "@wordpress/i18n";
+import * as React from "react";
+
+export const openEditModeDialog = (lockedBy: string) => {
```
Why is this function not being called?

MiryamOren commented on modules/components/module.php:
```diff
@@ -56,16 +67,121 @@ private function register_document_type( $documents_manager ) {
 			Component_Document::get_class_full_name()
 		);
-
-		register_post_type( Component_Document::TYPE, [
```
?

MiryamOren commented on modules/components/module.php:
```diff
@@ -56,16 +67,121 @@ private function register_document_type( $documents_manager ) {
+		$this->unlock_component( $post_id );
+	}
+
+	public function check_component_lock( $post_id ) {
```
Let's talk about it

elementorbot committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time

elementorbot committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time

MiryamOren commented on modules/components/module.php:
```diff
@@ -56,16 +74,121 @@ private function register_document_type( $documents_manager ) {
+	public function heartbeat_send( $data, $screen_id ) {
+		if ( 'elementor' === $screen_id && isset( $_GET['post'] ) ) {
+			$post_id = sanitize_text_field( wp_unslash( $_GET['post'] ) );
+			$data['elementor_component_lock'] = [
```
Don't we want to set this parameter only if the post is a component?

elementorbot committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time

elementorbot committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time

elementorbot committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time

elementorbot committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time

MuliElementor committed:
Refactor: Update document locking mechanism and REST API permissions

- Commented out unused lock and unlock routes in Components_REST_API.
- Changed permission callback for lock status route to allow users with 'edit_posts' capability.
- Introduced a new method to check if the current user is allowed to edit based on document lock status.
- Updated Document_Lock_Manager to use constants for lock metadata and simplified locking logic.
- Adjusted API response types to reflect changes in lock status handling.
- Enhanced component type creation to include callback for handling locked documents.

MuliElementor committed:
Refactor: Update document locking mechanism and REST API permissions

- Commented out unused lock and unlock routes in Components_REST_API.
- Changed permission callback for lock status route to allow users with 'edit_posts' capability.
- Introduced a new method to check if the current user is allowed to edit based on document lock status.
- Updated Document_Lock_Manager to use constants for lock metadata and simplified locking logic.
- Adjusted API response types to reflect changes in lock status handling.
- Enhanced component type creation to include callback for handling locked documents.

MuliElementor committed:
Merge branch 'ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time' of https://github.com/elementor/elementor into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time

MuliElementor committed:
add tests

MuliElementor committed:
add tests

MuliElementor committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time

MuliElementor committed:
update

MuliElementor committed:
update

MuliElementor committed:
Enhance: Implement lock and unlock functionality in Components REST API

- Added REST API routes for locking and unlocking components, with permission checks for users with 'edit_posts' capability.
- Implemented the lock_component and unlock_component methods to handle the locking logic.
- Updated tests to verify the new lock and unlock endpoints, including authentication and parameter validation checks.
- Removed commented-out code related to locking functionality for clarity.

MuliElementor committed:
Enhance: Implement lock and unlock functionality in Components REST API

- Added REST API routes for locking and unlocking components, with permission checks for users with 'edit_posts' capability.
- Implemented the lock_component and unlock_component methods to handle the locking logic.
- Updated tests to verify the new lock and unlock endpoints, including authentication and parameter validation checks.
- Removed commented-out code related to locking functionality for clarity.

MuliElementor committed:
Resolve merge conflict in components-rest-api.php

MuliElementor committed:
Resolve second merge conflict in components-rest-api.php

MuliElementor committed:
update

MuliElementor committed:
Merge branch 'ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time' of https://github.com/elementor/elementor into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time

MuliElementor committed:
format

MuliElementor committed:
fix tests

MuliElementor committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time

MiryamOren commented on packages/packages/core/editor-components/src/api.ts:
```diff
@@ -36,4 +45,49 @@ export const apiClient = {
+			.then( ( res ) => res.data );
+
+		return response.is_current_user_locked;
+	},
```
We don't use this method anywhere
```suggestion
```

MiryamOren commented on packages/packages/core/editor-components/src/api.ts:
```diff
@@ -36,4 +45,49 @@ export const apiClient = {
+			.then( ( res ) => res.data ),
+};
+
+export const canSwitchDocument = async (
```
Please do not export a separate function, but parse the data in `getComponentLockStatus` `then` cb, for example, see
`packages/packages/libs/editor-current-user/src/api.ts` line 22
```
	get: () =>
		httpService()
			.get< UserModel >( RESOURCE_URL, getUserPayload )
			.then( ( res ) => {
				const { capabilities = [], suppressedMessages = [] } = res.data;

				return { capabilities, suppressedMessages };
			} ),
```

MuliElementor committed:
Refactor: Update locking mechanism and API permissions for components

- Renamed lock manager methods for clarity and consistency.
- Changed permission callback for locking and unlocking routes to require 'manage_options' capability.
- Updated locking logic in the Components_REST_API to use the new lock component manager.
- Enhanced the Document_Lock_Manager to respect existing locks and return detailed lock status.
- Adjusted API client to handle updated lock status responses and modified component type creation to show locked user information.
- Updated tests to reflect changes in locking behavior and ensure consistency in lock metadata handling.

MuliElementor committed:
remove unused code

MuliElementor committed:
lint

MuliElementor committed:
update tests

MuliElementor committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time

MuliElementor committed:
lint

MuliElementor committed:
update

MuliElementor committed:
update

MuliElementor committed:
Refactor: Improve locking mechanism and API structure

- Introduced a singleton pattern for the Lock_Component_Manager to ensure a single instance is used throughout.
- Updated the Components_REST_API to cache the lock component manager instance for efficiency.
- Changed permission callback for API routes to require 'manage_options' capability.
- Cleaned up unused code and improved readability in the Lock_Component_Manager and Document_Lock_Manager classes.
- Enhanced the in-edit mode dialog for better user experience with improved formatting and comments.

MuliElementor committed:
update

MuliElementor committed:
Refactor: Standardize locking method names and improve test clarity

- Renamed methods in Lock_Component_Manager and Document_Lock_Manager from `lock_document` to `lock` for consistency.
- Updated corresponding test cases to reflect the new method names and improve readability.
- Enhanced test assertions to ensure accurate validation of locking behavior.

MuliElementor committed:
Enhance EditModeDialog: Extracted translation logic for clarity

- Moved translation logic for the editing notification into a separate constant for improved readability.
- Added a comment to clarify the purpose of the translation string for future translators.

MuliElementor committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time

MuliElementor committed:
lint

MuliElementor committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time
